### DISCUSSION TITLE
using PG INHERITS partitions without a discriminator in the ORM

### DISCUSSION THREAD

**Discussion by Alc-Alc** [sqlalchemy/sqlalchemy#9784] [Usage Questions] [answered]
How should I create my models such that the following DDL is emitted by SQLAlchemy?

[Taken from PostgreSQL docs](https://www.postgresql.org/docs/current/tutorial-inheritance.html)

```sql
CREATE TABLE cities (
  name       text,
  population real,
  elevation  int     -- (in ft)
);

CREATE TABLE capitals (
  state      char(2) UNIQUE NOT NULL
) INHERITS (cities);
```

I am able to get a similar DDL if I use `Table` (confirmed by looking at the logs)

```python
from sqlalchemy import create_engine, Table, Column, Text, REAL, Integer, String
from sqlalchemy.orm import DeclarativeBase

class Base(DeclarativeBase):
    pass

engine = create_engine("postgresql+psycopg://connection_string", echo=True)

cities = Table(
    "cities",
    Base.metadata,
    Column("name", Text),
    Column("population", REAL),
    Column("elevation", Integer),
)
capitals = Table(
    "capitals",
    Base.metadata,
    Column("state", String(2), nullable=False, unique=True),
    postgresql_inherits=cities.name,
)

Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)
```

How can I do the same using ORM? This is my attempt so far

```python
class City(Base):
    __tablename__ = 'cities'
    name: Mapped[str] = mapped_column(primary_key=True)
    population: Mapped[float]
    elevation: Mapped[int]


class Capital(City):
    __tablename__ = 'capitals'
    state: Mapped[str] = mapped_column(unique=True)
    __table_args__ = {"postgresql_inherits": City.__table__.name}
```

It raises an exception

> sqlalchemy.exc.NoForeignKeysError: Can't determine the inherit condition between inherited table 'cities' and inheriting table 'capitals'; tables have no foreign key relationships established.  Please ensure the inheriting table has a foreign key relationship to the inherited table, or provide an 'on clause' using the 'inherit_condition' mapper argument.

**Comment by zzzeek** [comment]
hi -

there are different approaches here.   We have a recipe for using PG INHERITS in such a way that the code would refer only to the "parent" table, "cities" in this case, using triggers as demonstrated in PG's docs at https://www.postgresql.org/docs/9.0/ddl-partitioning.html#DDL-PARTITIONING-IMPLEMENTATION .   This treats the subclasses as "single table inherits" classes that have no table of their own.

However, in your example you dont have any kind of "discriminator" column, and it implies that use of the "capitals" table directly is preferred.  To achieve this, we would have to give the ORM mapping various hints in how to access this table as well as what columns it would have, since the ORM mapping needs to know about these.    

First, if you have Capital extending from City, it's going to want to create  [joined table inheritance](https://docs.sqlalchemy.org/en/20/orm/inheritance.html#joined-table-inheritance) by default - that won't work here.  for PG inherits, if you are reading and writing to the "Capital" class, you can refer to the "capitals" table alone, it has all the columns of "cities" included.   so you would use [concrete inheritance](https://docs.sqlalchemy.org/en/20/orm/inheritance.html#concrete-table-inheritance) here.  however, then the ORM would need to know about all the columns that are present directly in "capitals".     So we have to do a bit of a hack here by enumerating to the ORM all the columns that this "capitals" table has, while omitting them from the DDL CREATE TABLE statement.  there's probably a bunch of ways to do this, but one way that is at least "simple" if not a little verbose is to enumerate the columns directly and omit them from CREATE TABLE by adding system=True to them.   The below POC illustrates a round trip for "capitals":

```py
from __future__ import annotations

from sqlalchemy import create_engine, select
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


engine = create_engine(
    "postgresql+psycopg://scott:tiger@localhost/test", echo=True
)


class City(Base):
    __tablename__ = "cities"
    name: Mapped[str] = mapped_column(primary_key=True)
    population: Mapped[float]
    elevation: Mapped[int]


class Capital(City):
    __tablename__ = "capitals"

    # this part can perhaps be automated by various means
    name: Mapped[str] = mapped_column(primary_key=True, system=True)
    population: Mapped[float] = mapped_column(system=True)
    elevation: Mapped[int] = mapped_column(system=True)

    state: Mapped[str] = mapped_column(unique=True)
    __mapper_args__ = {"concrete": True}
    __table_args__ = {"postgresql_inherits": City.__table__.name}


Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

s = Session(engine)

c1 = Capital(name="tiny", population=234065, elevation=100, state="NY")
s.add(c1)


c2 = s.scalars(select(Capital).where(Capital.state == "NY")).one()

assert c2 is c1

```

**Reply by zzzeek** [reply, marked as answer]
I've just updated https://github.com/sqlalchemy/sqlalchemy/wiki/PostgreSQLInheritance for modern patterns.     You might want to check this out because using PG INHERITS typically implies that it is doing transparent partitioning.  However you'd need to include a discriminator column.

**Reply by Alc-Alc** [reply]
Thanks for the response zzzeek. One of my searches led me to the linked recipe but I dismissed it considering it was from 2014 (on hindsight I shouldn't have, +1 for the update). Thanks again for the suggested alternative and I appreciate your work on SQLAlchemy 😇
